include/zephyr/bluetooth/classic/hfp_hf.h

```diff
@@ -168,16 +429,520 @@ struct bt_hfp_hf_cb {
+ *  of the AG.
+ *  Send the ATD>Nan... command to start memory dialing.
+ *  The result of the command will be notified through the callback
+ *  `dialing`.
+ *
+ *  @param hf HFP HF object.
+ *  @param location Memory location.
+ *
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_hfp_hf_memory_dial(struct bt_hfp_hf *hf, const char *location);
+
+/** @brief Handsfree HF redial last number
+ *
+ *  Initiate outgoing voice calls by recalling the last number
+ *  dialed by the AG.
+ *  Send the AT+BLDN command to recall the last number.
+ *  The result of the command will be notified through the callback
+ *  `dialing`.
+ *
+ *  @param hf HFP HF object.
+ *
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_hfp_hf_redial(struct bt_hfp_hf *hf);
+
+/** @brief Handsfree HF setup audio connection
+ *
+ *  Setup audio conenction by sending AT+BCC.
+ *  If @kconfig{CONFIG_BT_HFP_HF_CODEC_NEG} is not enabled, the error
+ *  `-ENOTSUP` will be returned if the function called.
+ *
+ *  @param hf HFP HF object.
+ *
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_hfp_hf_audio_connect(struct bt_hfp_hf *hf);
+
+/** @brief Handsfree HF set selected codec id
+ *
+ *  Set selected codec id by sending AT+BCS. The function is used to
+ *  response the codec negotiation request notified by callback
+ *  `codec_negotiate`. The parameter `codec_id` should be same as
+ *  `id` of callback `codec_negotiate` if the id could be supported.
+ *  Or, call `bt_hfp_hf_set_codecs` to notify the AG Codec IDs supported
+ *  by HFP HF.
+ *  If @kconfig{CONFIG_BT_HFP_HF_CODEC_NEG} is not enabled, the error
+ *  `-ENOTSUP` will be returned if the function called.
+ *
+ *  @param hf HFP HF object.
+ *  @param codec_id Selected codec id.
+ *
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_hfp_hf_select_codec(struct bt_hfp_hf *hf, uint8_t codec_id);
+
+/** @brief Handsfree HF set supported codec ids
+ *
+ *  Set supported codec ids by sending AT+BAC. This function is used
+ *  to notify AG the supported Codec IDs of HF.
+ *  If @kconfig{CONFIG_BT_HFP_HF_CODEC_NEG} is not enabled, the error
+ *  `-ENOTSUP` will be returned if the function called.
+ *
+ *  @param hf HFP HF object.
+ *  @param codec_ids Supported codec IDs.
+ *
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_hfp_hf_set_codecs(struct bt_hfp_hf *hf, uint8_t codec_ids);
+
+/** @brief Handsfree HF turns off AG's EC and NR
+ *
+ *  Turn off the AG's EC and NR by sending `AT+NREC=0`.
+ *  The result of the command is notified through the callback
+ *  `ecnr_turn_off`.
+ *  If @kconfig{CONFIG_BT_HFP_HF_ECNR} is not enabled, the error
+ *  `-ENOTSUP` will be returned if the function called.
+ *
+ *  @param hf HFP HF object.
+ *
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_hfp_hf_turn_off_ecnr(struct bt_hfp_hf *hf);
```

Comment:
Need bt_hfp_hf_turn_on_ecnr() api?

Reply:
From HFP_v1.9.pdf, the AG’s EC and NR is not needed to be turned off. Please refer to section `4.24 The HF Requests Turning off the AG’s EC and NR` of HFP specification.